sw/airborne/modules/pose_history/pose_history.c

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright (C) Roland Meertens
+ *
+ * This file is part of paparazzi
+ *
+ * paparazzi is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2, or (at your option)
+ * any later version.
+ *
+ * paparazzi is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with paparazzi; see the file COPYING.  If not, see
+ * <http://www.gnu.org/licenses/>.
+ */
+/**
+ * @file "modules/pose_history/pose_history.c"
+ * @author Roland Meertens
+ * Ask this module for the pose the drone had closest to a given timestamp
+ */
+
+#include "modules/pose_history/pose_history.h"
+#include <sys/time.h>
+#include "mcu_periph/sys_time.h"
+#include "state.h"
+
+#define POSE_HISTORY_SIZE 1024 // 1024 is two seconds
+
+typedef struct {
+  uint32_t timestamp;
+  struct FloatEulers rotation;
+} timeAndRotation;
+
+typedef struct {
+  uint32_t ring_index;
+  uint32_t ring_size;
+  timeAndRotation ring_data[POSE_HISTORY_SIZE];
+} rotationHistoryRingBuffer;
+
+rotationHistoryRingBuffer location_history;
+
+/**
+ * Given a pprz timestamp in used (obtained with get_sys_time_usec) we return the pose in FloatEulers closest to that time.
+ */
+struct FloatEulers get_eulers_at_timestamp(uint32_t timestamp)
+{
+  uint32_t index_history = 0;
+  uint32_t closestTimeDiff = abs(timestamp - location_history.ring_data[0].timestamp);
+  uint32_t closestIndex = 0;
+
+  // Seach for the timestamp closes to the timestamp argument of this function.
+  for (index_history = 0; index_history < location_history.ring_size; index_history++) {
```

Comment:
Not sure if your proposed function is better. The ring buffer is indeed ordered by time, but index zero can be way after our timestamp that we want to find... this means that we might increase our diff for a while before we jump back in time and are closer to the point we want to be at.